awsplugins/ec2/ec2.go

```diff
@@ -9,35 +9,115 @@
-	"github.com/aws/aws-sdk-go/aws/session"
+	"bytes"
+	"encoding/json"
+	"net/http"
+
 	"github.com/aws/aws-xray-sdk-go/internal/logger"
 	"github.com/aws/aws-xray-sdk-go/internal/plugins"
 )
 
 // Origin is the type of AWS resource that runs your application.
 const Origin = "AWS::EC2::Instance"
 
-// Init activates EC2Plugin at runtime.
+type Metdata struct {
+	AccountID               string
+	Architecture            string
+	AvailabilityZone        string
+	BillingProducts         string
+	DevpayProductCodes      string
+	MarketplaceProductCodes string
+	ImageID                 string
+	InstanceID              string
+	InstanceType            string
+	KernelID                string
+	PendingTime             string
+	PrivateIP               string
+	RamdiskID               string
+	Region                  string
+	Version                 string
+}
+
+//Init activates EC2Plugin at runtime.
 func Init() {
 	if plugins.InstancePluginMetadata != nil && plugins.InstancePluginMetadata.EC2Metadata == nil {
 		addPluginMetadata(plugins.InstancePluginMetadata)
 	}
 }
 
 func addPluginMetadata(pluginmd *plugins.PluginMetadata) {
-	session, e := session.NewSession()
-	if e != nil {
-		logger.Errorf("Unable to create a new ec2 session: %v", e)
-		return
+	var instanceData Metdata
+	imdsURL := "http://169.254.169.254/latest/"
+
+	client := &http.Client{
+		Transport: http.DefaultTransport,
+	}
+
+	token, err := getToken(imdsURL, client)
+	if err != nil {
+		logger.Errorf("Unable to fetch EC2 instance metadata token fallback to IMDS V1: %v", err)
 	}
-	client := ec2metadata.New(session)
-	doc, err := client.GetInstanceIdentityDocument()
+
+	resp, err := getMetadata(imdsURL, client, token)
 	if err != nil {
 		logger.Errorf("Unable to read EC2 instance metadata: %v", err)
 		return
 	}
 
-	pluginmd.EC2Metadata = &plugins.EC2Metadata{InstanceID: doc.InstanceID, AvailabilityZone: doc.AvailabilityZone}
+	buf := new(bytes.Buffer)
+	if _, err := buf.ReadFrom(resp.Body); err != nil {
+		logger.Errorf("Error while reading data from response buffer: %v", err)
+	}
+	metadata := buf.String()
+
+	if err := json.Unmarshal([]byte(metadata), &instanceData); err != nil {
+		logger.Errorf("Error while unmarshal operation: %v", err)
+	}
+
+	pluginmd.EC2Metadata = &plugins.EC2Metadata{InstanceID: instanceData.InstanceID, AvailabilityZone: instanceData.AvailabilityZone}
 	pluginmd.Origin = Origin
 }
+
+// getToken fetches token to fetch EC2 metadata
+func getToken(imdsURL string, client *http.Client) (string, error) {
+	ttlHeader := "X-aws-ec2-metadata-token-ttl-seconds"
+	defaultTTL := "60"
+	tokenURL := imdsURL + "api/token"
+
+	req, _ := http.NewRequest("PUT", tokenURL, nil)
+	req.Header.Add(ttlHeader, defaultTTL)
+	resp, err := client.Do(req)
+	if err != nil {
+		return "fallback", err
+	}
+
+	buf := new(bytes.Buffer)
+	if _, err := buf.ReadFrom(resp.Body); err != nil {
+		logger.Errorf("Error while reading data from response buffer: %v", err)
```

Comment:
yes but this would give false impression because this is not token retrieve failure. this is actually parsing failure. we don't want to fallback to IMDSv1 in the case of unable to parse token data right ?

Reply:
Well if we don't return here what will happen? It will return a corrupted string token that the plugin will attempt to use for a V2 request, which could cause unexpected behavior. 

I think returning the `"fallback", err` is acceptable here since we already have an invalid token so we know V2 won't work. There's no harm in just ignoring the token and trying V1 IMO, worst case we just log another failure if V1 doesn't work either.

Reply:
Agree that if the token can't be read (seems like it'd be for a connection reset when reading a very short string, so very unlikely but yeah can happen), 

- Optionally retry token fetch with upper limit / backoff
- Send request without token and it will fail if V1 is disabled but succeed otherwise

Sending a bad token should cause the metadata fetch to be guaranteed to fail, so there's no point to do it, we should either fallback, retry, or give up in our code.

Reply:
Okay. Makse sense! I will return an empty string here too.  so that it fallbacks to v1.